--- EntityId.gen	2011-07-15 15:37:23.000000000 -0700
+++ EntityID.java	2011-07-15 15:39:14.000000000 -0700
@@ -203,4 +203,21 @@
 
     return ivarsEqual;
  }
+
+/**
+ * Override of base class hashCode. This is convienent to use when looking up entities in
+ * a hash table. It uses the application and entity IDs to fill out the 32 bits. The idea
+ * is to get a "pretty close" to unique hashcode for each entity ID in 32 bits, and the
+ * best way to do that is leave out the site ID while keeping the rest, which tends to
+ * be more unique.
+ */
+ @Override
+ public int hashCode()
+ {
+     int hashcode = application; 
+     hashcode = hashcode << 16;
+     hashcode = hashcode + entity;
+     return hashcode;
+ }
+
 } // end of class
